### PR TITLE
added support for tcp proxyProtocol v1&v2 to backend

### LINKED ISSUE #5371: Send proxy protocol information to service/backend.
### Do you want to request a *feature* or report a *bug*?

Feature

### What did you expect to see?

Traefik in V2 can **accept** proxyprotocol (https://docs.traefik.io/v2.0/routing/entrypoints/#proxyprotocol).
It would be nice, if it could **send** proxyprotocol headers to service/backend.
As it now support **TCP** loadbalancer, we could transmit original IP to the service/backend.

Actually, there is no options for this. We tried to set an apache backend with `RemoteIPProxyProtocol On` (https://httpd.apache.org/docs/2.4/mod/mod_remoteip.html#remoteipproxyprotocol), but it don't work (server errors) because Traefik don't send proxyprotocol headers (it is required for apache if this setting is activated).

Purpose of this new feature is to be able to use TCP for HTTPS loadbalancing which will transmit original IP. It will make faster/direct HTTPS request  and there will be no more need to have (X-... headers).

I hope this proposition of new feature make sense.

Thanks.

### PR DESCRIPTION
**Pull request by mschneider82**
### What does this PR do?

added a "proxyProtocol" struct with a version setting to loadbalancer config to allow injecting the proxyprotocol v1 or v2 headers to a tcp backend connection. updated crds

"" disable (default)
"1" use proxy protocol version1 (human readable format)
"2" use proxy protocol version2

with version 2 its possible to also add tls infos. but in this implementation only the source ip+port and dst ip+port is added.

### What else does this PR do?

we also change the Listenercode to only use one proxyprotocol library: dropped c0va23/go-proxyprotocol for pires/go-proxyproto 

### Motivation

Currently there is no backend proxy protocol support, without this its not possible to forward smtp via tcp without loosing the client IP (its needed for detecting spam e.g.), with this postfix can be used in kubernetes with traefik as ingress tcp loadbalancer.

Fixes #5371 

### More

- [x] Added/updated tests 
- [x] Added/updated documentation

### CODE REVIEW

**Comment by mschneider82** [comment]
the alternative "config" solution would be to change the setting from boolean to "proxyProtocolVersion int" 

0 disable (default)
1 could be proxy protocol version1
2 could be proxy protocol version2

the current implementation does not require external libs. to support version 1&2 i would use github.com/pires/go-proxyproto
If this gets accepted i can implement this.

Adding TLS/SSL infos to the v2 impl. is complicated because i didnt see a way to get the TLS informations

**Comment by mschneider82** [comment]
Its done, please review

**Comment by mschneider82** [comment]
@ldez do you have time to review this? Its a long wanted feature

**Comment by ldez** [comment]
Hello,

Your PR has been classified with the `status/1-needs-design-review`:  this label is assigned on PR with a scope of impact that we considered as significant.

So it means that the review cycle will be longer than a more "basic" PR.

FYI, at this stage, before the evaluation by the team, a PR can be refused because of design considerations.

https://doc.traefik.io/traefik/v2.3/contributing/submitting-pull-requests/#pr-content
https://doc.traefik.io/traefik/v2.3/contributing/maintainers/#status-workflow

**Comment by jspdown** [comment]
@mschneider82  Thanks for your work on this PR!

As said in the original issue, Traefik supports being behind a proxy using the Proxy protocol. This behavior, as you probably seen, is configurable on the entrypoints and we are using the [c0va23/go-proxyprotocol](https://github.com/c0va23/go-proxyprotocol) package for providing a Proxy protocol capable listener.

You did a great job on this PR but we would like to keep only one Proxy protocol dependency. That being said, your chosen dependency [pires/go-proxyproto](https://github.com/pires/go-proxyproto) would be suitable for all our needs as it supports both client and server side. After reviewing it, it appears to be a great choice from a spec compliance and extensibility perspective.

Could you update the Proxy protocol listener? https://github.com/traefik/traefik/blob/b5198e63c446efdb488abd2bea826a59f75b1d63/pkg/server/server_entrypoint_tcp.go#L358-L385

Some changes will be required for replacing the package. [pires/go-proxyproto](https://github.com/pires/go-proxyproto) supports [Policies](https://pkg.go.dev/github.com/pires/go-proxyproto#PolicyFunc) which is a good replacement for the `WithSourceChecker(sourceCheck)`.

**Comment by mschneider82** [comment]
I thought about changing the listener too, but I didn't want to make this PR too big (its already not small). If it's ok for you i will do it here

**Comment by jspdown** [comment]
@mschneider82 Yes it would be great. Thanks again for your time. Don't hesitate if you need some help.

**Comment by citizenken** [comment]
I think this is what we need for our use-case, but want to make sure. We currently have Traefik sitting behind an AWS Classic LB with ProxyProtocol v2 enabled. With the `--entryPoints.web.proxyProtocol.insecure` CLI option in testing, we're not seeing the client IP passing through to our service with a basic HTTP request.

It sounds like this config will add a header exposing that IP to downstream services. Is that correct?

**Comment by jspdown** [comment]
@citizenken Yes exactly.

File: pkg/config/dynamic/tcp_config.go
Comment (jspdown) [inline pkg/config/dynamic/tcp_config.go:77]: Could you replace the `ProxyProtocolVersion` type by a string? Having an integers as a version can be harmful if future versions doesn't follow this pattern. There'll be no need for a pointer here, its zero value would simply mean to not enable proxy protocol. And for the versions, `"1"` and `"2"` will be fine.

Comment (mschneider82) [reply pkg/config/dynamic/tcp_config.go:77]: my first solution had no pointer and default value 0, this resulted in changing so many test and integration files that i dropped this solution after changing so many lines of testcode. Followed your guide that the change should be reviewable.

Are you sure about string? There are two other *int values in same struct. 
In v2 spec the version size (the 13th byte) is only one byte

**Comment by mschneider82** [comment]
@jspdown listener done. unfortunately the CI tests are only OK after merge with master (because of a timezone change)

**Comment by ldez** [comment]
> unfortunately the CI tests are only OK after merge with master (because of a timezone change)

@mschneider82 you just have to rebase your PR.

File: pkg/config/dynamic/tcp_config.go
Comment (mschneider82) [reply pkg/config/dynamic/tcp_config.go:77]: please advice if it should be a string or int, i really recommend to keep the pointer.

Comment (jspdown) [reply pkg/config/dynamic/tcp_config.go]: I tried to replace the `*string` by a `string` and I didn't found any problems with the tests. We would definitely prefer a `string` as a pointer is not useful in this case.

Comment (mschneider82) [reply pkg/config/dynamic/tcp_config.go]: Ok I will try tomorrow. Maybe the int was the problem.

**Review by jspdown — CHANGES_REQUESTED** [review]
Hello @mschneider82 

Thanks for your work. I made some comments and suggestions.
And also, could you:
- add the new field in the dynamic references in the doc contents (file.toml, file.yml, marathon-labels.json and docker-labels.yml) and generate afterward the KV references, trough a go generate
- run `make generate-crd`

File: pkg/config/dynamic/tcp_config.go
Comment (mschneider82) [reply pkg/config/dynamic/tcp_config.go]: fixed in 2d85905

File: pkg/provider/kubernetes/crd/kubernetes_tcp.go
Comment (mschneider82) [reply pkg/provider/kubernetes/crd/kubernetes_tcp.go]: fixed in c2a1155

File: pkg/server/server_entrypoint_tcp.go
Comment (mschneider82) [reply pkg/server/server_entrypoint_tcp.go:362]: I can change this if its wanted, but then debug logging is lost. Keep or change?

File: pkg/server/server_entrypoint_tcp.go
Comment (mschneider82) [reply pkg/server/server_entrypoint_tcp.go]: fixed typo ff2719c

File: pkg/tcp/proxy.go
Comment (mschneider82) [reply pkg/tcp/proxy.go]: fixed in c0d4b89

**Comment by ldez** [comment]
@mschneider82 to apply the suggestion you can:
- use the "Commit suggestion" button
- or in the "Files changed" view, use the "Add suggestion to batch"

File: pkg/tcp/proxy.go
Comment (mschneider82) [reply pkg/tcp/proxy.go:55]: added validation with logging 031197e

File: docs/content/routing/services/index.md
Comment (mschneider82) [reply docs/content/routing/services/index.md]: fixed in 5e0a2af

File: docs/content/routing/services/index.md
Comment (mschneider82) [reply docs/content/routing/services/index.md]: fixed in 5e0a2af

File: docs/content/routing/services/index.md
Comment (mschneider82) [reply docs/content/routing/services/index.md]: fixed in 5e0a2af

File: docs/content/routing/services/index.md
Comment (mschneider82) [reply docs/content/routing/services/index.md]: removed servers in 5e0a2af

**Comment by mschneider82** [comment]
> @mschneider82 to apply the suggestions you can:
> 
> * use the "Commit suggestion" button
> * or in the "Files changed" view, use the "Add suggestion to batch"

sorry did not knew this feature... i have added commit hashes to the suggestions

File: pkg/server/server_entrypoint_tcp.go
Comment (jspdown) [reply pkg/server/server_entrypoint_tcp.go:362]: I missed this log. Yes you are right just keep it like that, we really need this log.

File: pkg/tcp/proxy_test.go
Comment (mschneider82) [reply pkg/tcp/proxy_test.go:85]: i have made tests stronger in e9b482a please check

**Comment by mschneider82** [comment]
is it normal that the ci tests randomly fails? (i had to force push the same commit 3 times today..)

**Comment by mschneider82** [comment]
i get errors when i run `make generate-crd`

**Comment by jspdown** [comment]
@mschneider82 Sadly yes some integration tests are still a bit flaky. I'm investigating on the failing test to see if I can spot a reason for this flakiness. 
About the `make generate-crd`, In order to work, traefik must be clone in your go path under `$GOPATH/src/github.com/traefik/traefik` and your `GOPATH` must be set. 
If the problem persists, could you paste the output of this command here?

**Review by mschneider82 — COMMENTED** [review]
just wondering why accesscontrol was added to zz_generated.deepcopy.go - seems that b5198e63c446efdb488abd2bea826a59f75b1d63 missed the generate

**Comment by jspdown** [comment]
@mschneider82  Apparently the `make generate-crd` hasn't been run on this commit. We will make sure to fix to keep a clean diff. Thanks for spotting this :+1:

**Comment by mschneider82** [comment]
i have removed this commit from here so we dont get merge conflics.

**Review by jspdown — CHANGES_REQUESTED** [review]
I spent some time testing this feature. It works great :+1: Just two last minor suggestions

File: pkg/server/service/tcp/service.go
Comment (ldez) [reply pkg/server/service/tcp/service.go:56]: I would prefer the `switch`

File: pkg/tcp/proxy.go
Comment (ldez) [reply pkg/tcp/proxy.go:47]: I would prefer the `switch`

File: pkg/config/dynamic/tcp_config.go
Comment (jspdown) [inline pkg/config/dynamic/tcp_config.go:76]: After discussing with the rest of the team, we think the attribute `ProxyProtocolVersion` might be misleading. Since it contains "version" in its name people might understand it has: 
> Proxy protocol is always enabled, `ProxyProtocolVersion` is just here for overriding the default version. 

To disambiguate the naming we suggest to change this field into a struct like this:

```suggestion
	ProxyProtocol *ProxyProtocol      `json:"proxyProtocol,omitempty" toml:"proxyProtocol,omitempty" yaml:"proxyProtocol,omitempty"`
```

And the `ProxyProtocol` struct would be defined as:

```go
type ProxyProtocol struct {
  Version string
}
```

Doing this would lead people to think, proxy protocol is disabled by default. This way we can even define a `SetDefaults` for the `ProxyProtocol` type to set its version to "2" by default. Then if the protocol evolves in the future and allow for more configuration than just the version, we will be already prepared.

**Review by jspdown — CHANGES_REQUESTED** [review]

File: pkg/config/dynamic/tcp_config.go
Comment (mschneider82) [reply pkg/config/dynamic/tcp_config.go:76]: Just a question:

```
// SetDefaults Default values for a TCPServersLoadBalancer.
func (l *TCPServersLoadBalancer) SetDefaults() {
	defaultTerminationDelay := 100 // in milliseconds
	l.TerminationDelay = &defaultTerminationDelay
	l.ProxyProtocol = &ProxyProtocol{Version: "2"}
}
```

this would enable proxyprotocol v2 by default? It should be disabled by default

Comment (ldez) [reply pkg/config/dynamic/tcp_config.go:76]:
```go
func (p *ProxyProtocol) SetDefaults() {}
```

File: pkg/server/service/tcp/service_test.go
Comment (jspdown) [inline pkg/server/service/tcp/service_test.go:175]:
```suggestion
			desc:        "Server with unknown ProxyProtocol version",
```

File: pkg/config/dynamic/tcp_config.go
Comment (mschneider82) [reply pkg/config/dynamic/tcp_config.go]: Where should ProxyProtocol.SetDefaults() be executed? I defined the function but its not called yet.

Comment (jspdown) [reply pkg/config/dynamic/tcp_config.go]: `SetDefaults` is executed automatically when the provider constructs the dynamic configuration. This is done when providers call `parser.Fill` from the `traefik/paerser` package. You can find more information on 
https://github.com/traefik/paerser/blob/d69e93b8aaa12317e7e8f609b490232d8ad5f175/parser/element_fill.go#L104

**Comment by jspdown** [comment]
@mschneider82 Could you rebase on master? Integration tests fail because of an issue with the haproxy image. It has been fixed in https://github.com/traefik/traefik/commit/b3d298ea557e1d285e3f569895dfbbce3bde438e

**Comment by mschneider82** [comment]
can i retrigger the CI via bot?

File: pkg/provider/kubernetes/crd/traefik/v1alpha1/ingressroutetcp.go
Comment (mschneider82) [inline pkg/provider/kubernetes/crd/traefik/v1alpha1/ingressroutetcp.go:62]: is the dependecy to dynamic ok here?

Comment (jspdown) [reply pkg/provider/kubernetes/crd/traefik/v1alpha1/ingressroutetcp.go:62]: Yes it's fine here. We are already having such dependencies in `pkg/provider/kubernetes/crd/traefik/v1alpha1/ingressroute.go` for the `dynamic.ResponseForwarding`.

Comment (ldez) [reply pkg/provider/kubernetes/crd/traefik/v1alpha1/ingressroutetcp.go:62]: yes

**Comment by mschneider82** [comment]
will this feature merged in v2.4 ?

**Comment by jspdown** [comment]
@mschneider82 Yes we are planning to release it in the v2.4.

**Review by jspdown — APPROVED** [review]
LGTM :+1:

**Review by rtribotte — APPROVED** [review]
Thanks 👍

**Review by ldez — APPROVED** [review]
Thanks :+1:

**Review by kevinpollet — APPROVED** [review]
LGTM 👍

**Review by jbdoumenjou — APPROVED** [review]
LGTM :+1:

**Comment by loreleimccollum-work** [comment]
Is there a release cycle for when the v2.4 will be out? Eagerly awaiting this fix in this PR!

**Comment by mschneider82** [comment]
good choice to change version back to int

**Comment by loreleimccollum-work** [comment]
@jspdown do you know when the 2.4 version will be released, we are eagerly awaiting this fix

**Comment by jspdown** [comment]
Hello @loreleimccollum-work It's going to be released in the coming weeks.

**Comment by loreleimccollum-work** [comment]
We just tried to apply the 2.4 rc1 to our cluster and are using an pod that can print out the client IP, and we are not seeing the client IP preserved with this 2.4 rc1. I was wondering if we are missing something in the setup. I looked at the docs on this PR, but didn't see much in way of making sure this is setup. We have this annotation set on the traefik service `service.beta.kubernetes.io/aws-load-balancer-proxy-protocol: *` and args on the traefik pod have       `--entryPoints.web.proxyProtocol.insecure
      --entryPoints.websecure.proxyProtocol.insecure` any suggestions?

**Pull request merged**
